File: solutions/python/grade-school/1/grade_school.py
class Student:
    def __init__(self, name: str, grade: int) -> None:
        self.name = name
        self.grade = grade
        self.is_added = True

    def __eq__(self, other: 'Student') -> bool:
        return self.name == other.name


class School:
    def __init__(self):
        self.students: list[Student] = []

    def add_student(self, name: str, grade: int) -> None:
        new_student = Student(name, grade)
        if new_student in self.students:
            new_student.is_added = False

        self.students.append(new_student)

    def roster(self):
        return [student.name for student in self.sorted_students if student.is_added]

    def grade(self, grade_number: int):
        return [student.name for student in self.sorted_students if student.grade == grade_number and student.is_added]

    def added(self):
        return [student.is_added for student in self.students]

    @property
    def sorted_students(self) -> list[Student]:
        return sorted(self.students, key=lambda student: (student.grade, student.name))

File: solutions/ruby/grade-school/1/grade_school.rb
class School
  class Student
    attr_accessor :name, :grade

    def initialize(name, grade)
      @name = name
      @grade = grade
    end
  end

  def initialize
    @students = []
  end

  def add(name, grade)
    new_student = !(roster.include? name)
    @students << Student.new(name, grade) if new_student
    new_student
  end

  def roster
    sorted_student.map(&:name)
  end

  def grade(grade)
    sorted_student.select{|student| student.grade == grade}.map(&:name)
  end

  private
  def sorted_student
    @students.sort_by{|student| [student.grade, student.name]}
  end
end